File: php/src/BinarySearchTree.php
<?php


namespace Tree;


class BinarySearchTree
{
    private ?Node $root = null;

    public function getRoot(): ?Node
    {
        return $this->root;
    }

    public function add($data)
    {
        if (! $this->root)
        {
            $this->root = new Node($data);
            return;
        }

        $this->innerAdd($this->root, $data);
    }

    private function innerAdd(Node $currentNode, $value)
    {
        if ($value > $currentNode->value)
        {
            if ($currentNode->right !== null)
            {
                $this->innerAdd($currentNode->right, $value);
                return;
            }
            $currentNode->right = new Node($value);
        }
        else
        {

            if ($currentNode->left !== null)
            {
                $this->innerAdd($currentNode->left, $value);
                return;
            }
            $currentNode->left = new Node($value);
        }
    }

    public function find(int $value): int
    {
        $node = $this->root;
        if (! $node)
        {
            return -1;
        }
        $iteration = 0;
        while (true)
        {
            if ($value === $node->value)
            {
                return $iteration;
            }
            if ($value > $node->value && $node->right)
            {
                $node = $node->right;
                $iteration++;
                continue;
            }
            else
            {
                if ($node->left !== null)
                {
                    $node = $node->left;
                    $iteration++;
                    continue;
                }
            }

            if (! $node->left && ! $node->right)
            {
                break;
            }
        }
        return -1;
    }
}

File: python3/binary_search_tree.py
from node import Node


class BinarySearchTree:
    __root: Node = None

    @property
    def root(self) -> Node:
        return self.__root

    def add(self, data):
        if self.root is None:
            self.__root = Node(data)
            return

        self.__add(self.__root, data)

    def find(self, data):
        iteration = 0

        if self.__root is None:
            return -1

        node = self.__root
        while True:

            if data == node.data:
                return iteration

            if node.right is None and node.left is None:
                break

            if data > node.data and node.right is not None:
                node = node.right
                iteration += 1
                continue

            else:
                if node.left is not None:
                    node = node.left
                    iteration += 1
                    continue

        return -1

    def __add(self, node: Node, data):

        if data > node.data:
            if node.right is not None:
                self.__add(node.right, data)
                return
            node.right = Node(data)
        else:
            if node.left is not None:
                self.__add(node.left, data)
                return
            node.left = Node(data)
